### PR TITLE
Replace boost::filesystem with std::filesystem

### PR DESCRIPTION
**Pull request by kiminuo**
## Introduction

C++17 has introduced the filesystem library. [cppreference.com](https://en.cppreference.com/w/cpp/filesystem) describes very well the origin of the library:

> The filesystem library was originally developed as boost.filesystem, was published as the technical specification ISO/IEC TS 18822:2015, and finally merged to ISO C++ as of C++17. The boost implementation is currently available on more compilers and platforms than the C++17 library.

This draft PR was created to gain feedback and examine what would be necessary to switch from `boost::filesystem` to `std::filesystem` in bitcoin codebase.

Bitcoin codebase contains `src/fs.h` which is a wrapper for the [currently used filesystem library](https://github.com/bitcoin/bitcoin/blob/6762a627ecb89ba8d4ed81a049a5d802e6dd75c2/src/fs.h#L18). The first impression is that one may just replace `namespace fs = boost::filesystem;` with `namespace fs = std::filesystem;` and all will be good. Unfortunately, there are some differences between the [Boost's filesystem library](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html) and the [c++17 filesystem library](https://en.cppreference.com/w/cpp/filesystem). A nice summary by Davis Herring can be read [here](https://stackoverflow.com/questions/40899267/how-similar-are-boost-filesystem-and-the-c-standard-filesystem-library).

## Differences between filesystem implementations that affects Bitcoin code

1. `boost::filesystem::system_complete()` was renamed to `std::filesystem::absolute()` \[[source](https://stackoverflow.com/a/46271698)\]

2. [boost::filesystem::absolute()](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#absolute) has second `const path& base` parameter which was dropped in [std::filesystem::absolute](https://en.cppreference.com/w/cpp/filesystem/absolute) (see [reasoning](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0492r2.html#US78)).

     Each such instance like [this one](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/blockchain.cpp#L2284) can be fixed by changing:

     `fs::path path = fs::absolute(request.params[0].get_str(), GetDataDir());`

     to

    `fs::path path = fs::absolute(GetDataDir() / request.params[0].get_str());` where `/` is function call of `operator/` function.

     Note: Notably `path("foo") / "c:/bar"` yields `c:/bar`. However, that's how `boost::filesystem::absolute(p, base_dir)` behaves too.

3. [boost::filesystem::path::imbue()](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#path-imbue) is not present in `std::filesystem::path`.

4. [boost::filesystem::unique_path()](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#unique_path) is not present in `std::filesystem`.

5. [boost::filesystem::equivalent(path1, path2)](https://github.com/boostorg/filesystem/blob/812cb9bad3bcbb2161c7cd2c1132b51c4b29c4b9/src/operations.cpp#L1241) differs from [std::filesystem::equivalent(path1, path2)](https://github.com/llvm/llvm-project/blob/00b161b8971bc6d3cb55f13502288b8fe0dbaa42/libcxx/src/filesystem/operations.cpp#L917) as the later reports *not supported* error if one of the files does not exist whereas Boost returns `false`.
    * See also [cppreference](https://en.cppreference.com/w/cpp/filesystem/equivalent).

## Support for C++17 `<filesystem`>

* Clang 7+ (see [[Phoronix](https://www.phoronix.com/scan.php?page=news_item&px=Filesystem-Lands-LLVM-Libcpp)][[git commit](https://github.com/llvm-mirror/libcxx/commit/a0866c5fb5020d15c69deda94d32a7f982b88dc9)])
* GCC 8.1+ ([[GCC status page](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017)] [[git commit](https://github.com/gcc-mirror/gcc/commit/3b90ed62fb848046ed7ddef07df7c806e7f3fadb)])
  * GCC 8.x requires linking with `-lstdc++fs`.

## Issues & PRs that may be linked with this PR

https://github.com/bitcoin/bitcoin/issues/13103, https://github.com/bitcoin/bitcoin/pull/13787, https://github.com/google/leveldb/pull/760, https://github.com/bitcoin/bitcoin/pull/6093

## TODO

* [x] Remove Boost's filesystem from https://github.com/bitcoin/bitcoin/blob/master/configure.ac (help is welcome)
  * Relevant PR: https://github.com/bitcoin/bitcoin/pull/7302 & https://github.com/bitcoin/bitcoin/commit/76ac35f36d87078da62f95b4a1167ec296e37363
  * Done in #19667.
* [ ] [Am I missing `-lstdc++fs`](https://www.bfilipek.com/2019/05/boost-to-stdfs.html)?
* [ ] What about https://github.com/bitcoin/bitcoin/pull/19933?
* [ ] Address this change https://github.com/bitcoin/bitcoin/pull/19960
* [ ] `unicodepath` test in  `src/test/dbwrapper_tests.cpp` seems to require some UTF-8 locale.
  * See also https://github.com/bitcoin/bitcoin/blob/ae9ee5bdb18c5babbfa2875ab694df118d6a1c1b/src/util/system.cpp#L1249
  * https://github.com/microsoft/STL/blob/2aa944beba9af1d72537fd5e35d402c8b830a5dc/stl/src/filesystem.cpp#L266 - C++17 in Microsoft C++ standard library
  * https://docs.microsoft.com/en-us/cpp/standard-library/filesystem?view=msvc-160
  * https://docs.microsoft.com/en-us/cpp/standard-library/file-system-navigation?view=msvc-160

## Inspiration

* https://github.com/ceptontech/filesystem
* https://github.com/ceztko/u8filesystem

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: this will be nice to have in the future when we switch to C++17. Thanks for experimenting!

**Comment by MarcoFalke** [comment]
Would it make sense to split out the two `absolute()` commits, because they make sense even with C++11?

**Comment by kiminuo** [comment]
> Would it make sense to split out the two `absolute()` commits, because they make sense even with C++11?

I think so. Yes.

**Comment by laanwj** [comment]
Nice work.

File: src/fs.cpp [hunk 1]
```diff
@@ -31,6 +32,14 @@ FILE *fopen(const fs::path& p, const char *mode)
 #endif
 }
 
+fs::path unique_path() {
+    // TODO: Do not use Boost in this implementation
+    boost::filesystem::path p = boost::filesystem::unique_path();
+    fs::path p2{p.string()};
+
+    return p2;
```
Comment (MarcoFalke) [inline src/fs.cpp:40]: I think all that boost does it pick 64 bits of randomness? Something along the lines of 

```suggestion
    return g_insecure_rand_ctx_temp_path.rand256().ToString().substr(0, 16);
```

should be identical.

Comment (kiminuo) [reply src/fs.cpp:40]: Yes. I believe so too.

However, using this leads to a new circular dependency "fs -> random -> logging -> fs": https://travis-ci.com/github/kiminuo/bitcoin/jobs/360282950#L466 (https://github.com/kiminuo/bitcoin/commit/333ab6ad421f1b25bf66bd2557cd3d0e4b60e0f6)

`unique_path()` is used only twice but I don't know what is preferred solution here. Inline it? Move function somewhere else? Where?

_____
For other people who may read this: For [used](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) Boost 1.70.0:

* Source code: https://github.com/boostorg/filesystem/blob/e49775e8bbc4d8dbea5fbf6cf7c29f4d7bbd0c67/src/unique_path.cpp#L131
* Docs: https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#unique_path

Comment (MarcoFalke) [reply src/fs.cpp:40]: I think there are two options:

* Inline
* or create a new header, e.g. `src/util/fs.h`

**Comment by elichai** [comment]
Nice! Please ping me when it's ready for review :)

**Comment by kiminuo** [comment]
@theuni Hi, would you know how to remove this reference to Boost's filesystem: 

https://github.com/bitcoin/bitcoin/blob/2a784723f0c0e642353dc74ec6aef4d5f8345044/configure.ac#L1285

? 

It comes from #7302 (commit 76ac35f)

**Comment by elichai** [comment]
@kiminuo Now that #20413 is merged, are you planning on rebasing this and getting it ready for review?

**Comment by kiminuo** [comment]
> @kiminuo Now that #20413 is merged, are you planning on rebasing this and getting it ready for review?

I will get to it. However, @MarcoFalke mentioned on IRC briefly that `std::fs` is not available on some supported OSs even with C++17 enabled. I'm not sure I understood correctly. If true, it would delay this feature months or years.

**Comment by MarcoFalke** [comment]
Hi, catching up on the IRC log.

I just wanted to say that this is blocked on #20460 and there is nothing you can do about this in this pull. Fixing #20460 will need build system and gitian/guix, as well as CI changes. This would be too much to do here.

**Comment by fanquake** [comment]
Thanks for your work here so far @kiminuo. I've addressed the build system requirements, and combined this changes into #20744 (still attributed to you). However due to runtime and compiler requirements it's unlikely we'll be able to start using them just yet. I'm going to close this PR in favour of #20744 for now.

**Pull request closed without merging**